src/inventory_ui.cpp

```diff
@@ -108,308 +393,413 @@ bool inventory_column::is_selected( const inventory_entry &entry ) const
+        const int sort_rank = category.sort_rank + dist;
+        const item_category new_category( id, name, sort_rank );
+
+        categories.push_back( std::move( new_category ) );
+    } else {
+        const auto existing = find_cat_by_id( category.id );
+        if( existing != nullptr ) {
+            return existing;
+        }
+
+        categories.push_back( category );
+    }
+
+    return &categories.back();
+}
+
+void inventory_selector::add_item( inventory_column &target_column,
+                                   const std::shared_ptr<item_location> &location,
+                                   size_t stack_size, const item_category *custom_category )
+{
+    if( !preset.is_shown( *location ) ) {
+        return;
+    }
+
+    inventory_entry entry( location, stack_size, custom_category );
+
+    entry.enabled = preset.is_enabled( entry.get_location() );
+    entry.rank = preset.get_rank( entry.get_location() );
+
+    target_column.add_entry( entry );
+    on_entry_add( entry );
+
+    layout_is_valid = false;
+}
+
+void inventory_selector::add_custom_items( inventory_column &target_column,
+                                           const std::list<item>::iterator &from,
+                                           const std::list<item>::iterator &to,
+                                           const item_category &custom_category,
                                            const std::function<std::shared_ptr<item_location>( item * )> &locator )
 {
     const auto &stacks = restack_items( from, to );
+    const item_category *nat_category = nullptr;
 
     for( const auto &stack : stacks ) {
-        const auto &location = locator( stack.front() );
-        if( filter( *location ) ) {
-            const std::string name = trim( string_format( _( "%s %s" ), to_upper_case( title ).c_str(),
-                                                          direction_suffix( u.pos(), location->position() ).c_str() ) );
-            if( categories.empty() || categories.back().id != name ) {
-                categories.emplace_back( name, name, INT_MIN + int( categories.size() ) );
-            }
-            if( custom_column == nullptr ) {
-                custom_column.reset( new inventory_column() );
-            }
-            const long invlet = ( cur_custom_invlet <= max_custom_invlet ) ? cur_custom_invlet++ : '\0';
-            custom_column->add_entry( inventory_entry( location, stack.size(), &categories.back(), c_unset, invlet ) );
+        const auto location = locator( stack.front() );
+
+        if( nat_category == nullptr && preset.is_shown( *location ) ) {
+            nat_category = naturalize_category( custom_category, location->position() );
         }
+
+        add_item( target_column, location, stack.size(), nat_category );
+    }
+}
+
+void inventory_selector::add_character_items( Character &character )
+{
+    const auto make_char_loc = [ &character ]( item &it ) {
```

Comment:
You should use `visitable` here and then calls to `is_worn`. This avoids the dependency on the internal implementation of `Character` which maybe subject to future change.